packages/flutter/lib/src/cupertino/text_selection.dart

```diff
@@ -446,5 +452,675 @@ class _CupertinoTextSelectionControls extends TextSelectionControls {
+      child: FadeTransition(
+        opacity: _controller,
+        child: _CupertinoTextSelectionToolbarItems(
+          page: _page,
+          backButton: CupertinoButton(
+            borderRadius: null,
+            color: _kToolbarBackgroundColor,
+            minSize: _kToolbarHeight,
+            onPressed: _handlePreviousPage,
+            padding: arrowPadding,
+            pressedOpacity: 0.7,
+            child: const Text('◀', style: _kToolbarButtonFontStyle),
+          ),
+          dividerWidth: 1.0 / MediaQuery.of(context).devicePixelRatio,
+          nextButton: CupertinoButton(
+            borderRadius: null,
+            color: _kToolbarBackgroundColor,
+            minSize: _kToolbarHeight,
+            onPressed: _handleNextPage,
+            padding: arrowPadding,
+            pressedOpacity: 0.7,
+            child: const Text('▶', style: _kToolbarButtonFontStyle),
+          ),
+          nextButtonDisabled: CupertinoButton(
+            borderRadius: null,
+            color: _kToolbarBackgroundColor,
+            minSize: _kToolbarHeight,
+            onPressed: () {},
+            padding: arrowPadding,
+            pressedOpacity: 1.0,
+            child: const Text('▶', style: _kToolbarButtonDisabledFontStyle),
+          ),
+          children: widget.children,
+        ),
+      ),
+    );
+  }
+}
+
+// The custom RenderObjectWidget that, together with
+// _CupertinoTextSelectionToolbarItemsRenderBox and
+// _CupertinoTextSelectionToolbarItemsElement, paginates the menu items.
+class _CupertinoTextSelectionToolbarItems extends RenderObjectWidget {
+  _CupertinoTextSelectionToolbarItems({
+    Key key,
+    @required this.page,
+    @required this.children,
+    @required this.backButton,
+    @required this.dividerWidth,
+    @required this.nextButton,
+    @required this.nextButtonDisabled,
+  }) : assert(children != null),
+       assert(children.isNotEmpty),
+       assert(backButton != null),
+       assert(dividerWidth != null),
+       assert(nextButton != null),
+       assert(nextButtonDisabled != null),
+       assert(page != null),
+       super(key: key);
+
+  final Widget backButton;
+  final List<Widget> children;
+  final double dividerWidth;
+  final Widget nextButton;
+  final Widget nextButtonDisabled;
+  final int page;
+
+  @override
+  _CupertinoTextSelectionToolbarItemsRenderBox createRenderObject(BuildContext context) {
+    return _CupertinoTextSelectionToolbarItemsRenderBox(
+      dividerWidth: dividerWidth,
+      page: page,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _CupertinoTextSelectionToolbarItemsRenderBox renderObject) {
+    renderObject.page = page;
+  }
+
+  @override
+  _CupertinoTextSelectionToolbarItemsElement createElement() => _CupertinoTextSelectionToolbarItemsElement(this);
+}
+
+// The custom RenderObjectElement that helps paginate the menu items.
+class _CupertinoTextSelectionToolbarItemsElement extends RenderObjectElement {
```

Comment:
I tried extending MultiChildRenderObjectElement when I wrote this and the reason I decided to do it this way instead was that I sometimes needed to call the super method on RenderObjectElement without going through MultiChildRenderObjectElement.

An example is `forgetChild`.  For a child that exists in `slotToChild`, I need to remove the child, then I need to call RenderObjectElement's forgetChild without calling MultiChildRenderObjectElement's forgetChild.  Let me know if I'm wrong, but I thought that was a showstopper since I can't do `super.super.forgetChild`.

Maybe a mixin will work though!  I'll try that right now.

Reply:
What is wrong if  we call MultiChildRenderObjectElement's forgetchild?

Reply:
I experimented a bit and I think it would work if I did something like this.  Let me know what you think.

```dart
mixin ContainerRenderObjectMixin on RenderObjectElement {
  List<Element> _children;
  final Set<Element> _forgottenChildren = HashSet<Element>();

  ...

  // Rather than overriding forgetChild, I give it a new name.
  void forgetListChild(Element child) {
    assert(_children.contains(child));
    assert(!_forgottenChildren.contains(child));
    _forgottenChildren.add(child);
    super.forgetChild(child);
  }

  ...

}

// Then I use the mixin here in my code.
class _CupertinoTextSelectionToolbarItemsElement extends RenderObjectElement with ContainerRenderObjectMixin {

  ...

  @override
  void forgetChild(Element child) {
    // Handle children in slots like normal.
    if (childToSlot.containsKey(child)) {
      final _CupertinoTextSelectionToolbarItemsSlot slot = childToSlot[child];
      childToSlot.remove(child);
      slotToChild.remove(slot);
      // Super calls to RenderObjectElement.forgetChild as desired.
      super.forgetChild(child);

    // Handle list children with the mixin, which can also call to RenderObjectElement.forgetChild.
    } else {
      forgetListChild(child);
    }
  }

  ...

}

// And we can also use the mixin in MultiChildRenderObjectElement to deduplicate code.
class MultiChildRenderObjectElement extends RenderObjectElement with ContainerRenderObjectMixin {

  ...

  @override
  void forgetChild(Element child) {
    forgetListChild(child);
  }

  ...

}
```

Reply:
@chunhtai Sorry, I missed your question because I didn't refresh.  Calling MultiChildRenderObjectElement's forgetChild with a slotted child will add it to _forgottenChildren, which is only used for list children.

I did work around this in the mixin I pasted above, though.